Generate the restraint

xiki-tempula opened the pull request:
Implement the ABFE restraint search routine.

Bug fix (non-breaking change which fixes an issue)

- Fix the behaviour intramol
- Find the version of the conda Gromacs.
- Fix the issue that velocity is not generated in equilibration protocol. 
- Takes in temperature as float
- The restraint object now takes in the temperature attribute

New feature (non-breaking change which adds functionality)
- Implement the ABFE restraint search routine.
- Add _FreeEnergyMixin._check_column_name as a convenient way of checking lambda.
- Add MDRestraintsGenerator as dependency

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -0,0 +1,939 @@
+import subprocess as _subprocess
+import tempfile as _tempfile
+import warnings as _warnings
+import zipfile as _zipfile
```
Lots of unused imports, so best to optimise them away I think

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -0,0 +1,939 @@
+        # Initialise the process.
+        self._initialise_process(self._system, gpu_support)
+
+    def start(self):
```
A lot of these functions seem to be directly taken from `Process`?

xiki-tempula replied:
They are created to mimic the BSS.FreeEnergy.Relative and is meant to be a high level class that the user directly interact with.

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -0,0 +1,939 @@
+        return self._work_dir
+
+    @staticmethod
+    def analyse(work_dir, system, traj, temperature, rest_type='Boresch',
```
Unless there's a good reason not to, I'd completely abstract this function away. It's big enough to warrant its own class, and can be used with an input `Process`, which can then be checked for everything you need, so you don't need to couple it to the `Process`.

xiki-tempula replied:
This class is created to provide a convenient way of accessing this function.
So two ways are provided to access this function.

- The experienced user could run the process and provide the trajectory as input to this staticmethod to do the analysis.
- Or the naive user provide protocol and this class take care of the rest in an end-to-end fashion.

It could be the case that the user generate a protocol, then use protocol to generate process, then feed the process to the class, but in this case, the user might just do one more step and use the analysis function as staticmethod.

msuruzhon replied:
Yes this is fine, the original `RestraintSearch` class then can still have this (let's call it `RestraintGenerator`) as a wrapped class which is called by this `analyse()` function. But in any case I think it makes sense to pull this out of `RestraintSearch` altogether and modularise properly.

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -0,0 +1,939 @@
+                else:
+                    raise ImportError('MDRestraintsGenerator not available.')
+            # No need to review this part as Finlay is still working on it.
+            elif method == "BSS":
```
If this is still work in progress, then shouldn't we remove this whole part and let him submit as a PR separately?

xiki-tempula replied:
I ask him to rest to this branch and develop based on that so I'm not sure if I shall just remove his code. I guess I could pack it into a function.

msuruzhon commented on test/Sandpit/Exscientia/Align/test_decouple.py:
```diff
@@ -29,7 +29,7 @@ def test_set_property_map(mol, charge0, charge1, LJ0, LJ1):
     assert new._sire_object.property('LJ0').value() == LJ0
     assert new._sire_object.property('LJ1').value() == LJ1
     assert new.isDecoupled() == True
-    assert new._sire_object.property('annihilated').value() == False
```
Don't `bool`s have to be compared with `a is True` rather than `a == True`? Otherwise `a` could not even be `bool` and it would still pass.

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -0,0 +1,939 @@
+
+    @staticmethod
+    def analyse(work_dir, system, traj, temperature, rest_type='Boresch',
+                method='BSS',         
```
Why is `BSS` the default if it is work in progress?

msuruzhon reviewed (changes requested):
Thanks @xiki-tempula I have left a few comments and questions. Also could we run CI just to make sure we haven't broken anything? For some reason it's not running.

xiki-tempula commented:
@msuruzhon Thanks. I will address the comments now. The CI is running?

msuruzhon commented:
Ah never mind, I think I got confused with the `This branch has not been deployed` message.

xiki-tempula commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -0,0 +1,939 @@
+if _sys.platform == "win32":
+    _analyse_freenrg = "%s %s" % (_os.path.join(_os.path.normpath(_getBinDir()), "sire_python.exe"), _analyse_freenrg)
+
+class RestraintSearch():
```
I has this thought in the beginning. Then it is kind of agreed that we will make a fresh class separate from the process such that in the sandpit to dev process, the layout could be consolidated.

xiki-tempula commented:
@msuruzhon thanks for the review. I have addressed all the comments. I have created the new class RestartAnalysis and put finaly's stuff at the very bottom.

msuruzhon reviewed (approved):
Thanks @xiki-tempula for the changes, you can merge once the checks pass.

xiki-tempula commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -0,0 +1,939 @@
+            recept_selection_str += ' and '
+            recept_selection_str += append_to_recept_selection
+
+        if rest_type.lower() == 'boresch':
```
@msuruzhon I have break down this function a bit so it is not a bit function. So at least the MDRestraintgeneration part is better.
